### PR TITLE
Added Android Support for Make display of snapped location configurab…

### PR DESCRIPTION
**Pull request by cu8code**
#209 

This PR adds the support for making the snapped location configurable by adding an additional variable `location: UserLocation`  that exposes the unsnapped location of the user!

### CODE REVIEW

File: android/core/src/main/java/com/stadiamaps/ferrostar/core/NavigationViewModel.kt
Comment (ianthetechie) [inline android/core/src/main/java/com/stadiamaps/ferrostar/core/NavigationViewModel.kt:69]: I don't think this is a valid default value. The location here is straight from the provider (unsnapped).

File: android/core/src/main/java/com/stadiamaps/ferrostar/core/NavigationViewModel.kt
Comment (ianthetechie) [inline android/core/src/main/java/com/stadiamaps/ferrostar/core/NavigationViewModel.kt:76]: I think I didn't specify very well how I envision this being used (my bad!).

I _think_ it would be useful to have a Boolean argument to the default navigation view model which controls whether the snapped location is used (when available).

It would still need to fall back to the unsnapped location when not navigating. Having a booked though to prefer snapped location would allow for as much flexibility as I can think of at the moment. Maybe @Archdoog also has an opinion on the most useful interface.

File: android/core/src/main/java/com/stadiamaps/ferrostar/core/NavigationViewModel.kt
Comment (ianthetechie) [inline android/core/src/main/java/com/stadiamaps/ferrostar/core/NavigationViewModel.kt:91]: Do we want to use the last location for both cases here? I think should be the unsnapped value and the other could be snapped.

**Review by ianthetechie — CHANGES_REQUESTED** [review]
Thanks for the PR! Few comments.... I may have missed some context though as I'm reviewing on mobile but wanted to give some initial feedback.

Also you can fix the build issues by running kftmt. There are details in CONTRIBUTING.md

File: android/core/src/main/java/com/stadiamaps/ferrostar/core/NavigationViewModel.kt
Comment (Archdoog) [reply android/core/src/main/java/com/stadiamaps/ferrostar/core/NavigationViewModel.kt:76]: Thinking through the goals here, maybe we should:

1. Add a `rawLocation` to the NavigationState in common. This would generally simplify the stack if a dev wants to use it for some custom behavior (whether UI or other). This is easy to accomplish since we already have single location provider under the hood passing through the `NavigationController`.
2. We could enhance this by providing an encapsulated `fun userLocation(userLocationMode: UserLocationMode) -> UserLocation`. Where `UserLocationMode` is an enum of raw location, snapped location, etc. Then the view model could just publish whatever the location is for the desired mode. This behavior/pattern could also be more useful globally for other platforms - not just android.

Side note, this might be a good point to brainstorm feature bloat on `NavigationState`. I think we're consistently going to have the desire to pack more information into that object (or multiple objects that distill navigation info), so we might want to standardize our process a bit. E.g. separate objects or namespaces like `navigationState.location`, `navigationState.instructions`, etc? 

Thoughts?

Comment (ianthetechie) [reply android/core/src/main/java/com/stadiamaps/ferrostar/core/NavigationViewModel.kt:76]: I feel like adding this to the `NavigationState` is a bad idea for several reasons (and said as much in this comment: https://github.com/stadiamaps/ferrostar/issues/209#issuecomment-2342570373).

* As you said, it is really bloating the state
* It doesn't even have anything to do with navigation. Snapped location is very much specific to navigation, but the user's "raw" location is not in any way tied to `NavigationState`. We want to use this outside the context of a trip; it's not hard to get the data; we just want to expose it in the observed _UI state_.

Comment (ianthetechie) [reply android/core/src/main/java/com/stadiamaps/ferrostar/core/NavigationViewModel.kt:76]: @Archdoog and I discussed this a bit in the core devs meeting today and will play around with it as-is first before making a decision on what changes need to be made (if any; we realized that it might actually be great as-is but that we just need to actually _use_ it first). No action needed on your part right now @cu8code; we'll probably either 1) merge as-is (after fixing lints or w/e), or 2) have a better idea after using it for a bit.

Sorry to hold this up, but I'll revisit this later in the week :)

Comment (ianthetechie) [reply android/core/src/main/java/com/stadiamaps/ferrostar/core/NavigationViewModel.kt:94]: Sorry for the delay on this. Yes, it looks like your existing approach is all we need at the view model level :) I've gone ahead and added the conveniences I was looking for at a higher level (the composable views), which is where they belong anyways.

**Review by ianthetechie — APPROVED** [review]

**Pull request merged**
